sros2/sros2/api/_key.py

```diff
@@ -103,12 +104,14 @@ def create_key(keystore_path, identity):
 def list_keys(keystore_path):
     enclaves_path = _keystore.get_keystore_enclaves_dir(keystore_path)
     if not os.path.isdir(keystore_path):
-        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), keystore_path)
+        raise FileNotFoundError(
+            errno.ENOENT, os.strerror(errno.ENOENT), keystore_path)
     if not os.path.isdir(enclaves_path):
         return True
-    for name in os.listdir(enclaves_path):
-        if os.path.isdir(os.path.join(enclaves_path, name)):
-            print(name)
+    p = pathlib.Path(enclaves_path)
+    key_file_paths = sorted(p.glob('**/key.pem'))
+    for key_file_path in key_file_paths:
+        print('/{}'.format(key_file_path.parent.relative_to(enclaves_path)))
```

Comment:
Should this use `os.sep`?

Reply:
If it uses the same / used in designating the enclave name path, then the stdout could be piped for more commands involving the enclave name.

Reply:
That's a good idea, @ruffsl. I actually think my initial comment is wrong; I believe pathlib will use forward slashes regardless of the OS, but your suggestion requires a consistency in path handling that I don't think we have right now. We should take a pass on converting things to pathlib and then we can use forward slashes reliably. That's outside this PR, of course.

Reply:
Running the code as-is on windows, with a file `.\foo\bar\key.pem', gave a printout `/foo\bar`

Reply:
Pathlib can read './foo\\bar` just fine and resolve it to the same path

Reply:
Thanks @Arnatious. I don't think we want to print `/foo\bar` though. I'm back to my original suggestion of using `os.sep`, or we can use `as_posix()` when printing them to stdout.

Reply:
Good point the current bewhavior is not what we'd want.

> I'm back to my original suggestion of using os.sep

It depends what we think the function returns.
If it is to give print a list of enclave names, the fully qualified name should be used (forward slashes regardless of the platform)
If it is to print a list of enclave paths it should us the one from the OS (and likely not use an absolute path but a relative one ?)

I'm personally leaning towards the former (return the fully qualified name that's used in the certs/permission files)

Reply:
> I'm personally leaning towards the former (return the fully qualified name that's used in the certs/permission files)

Agreed. `as_posix()`, then?

Reply:
> Agreed. as_posix(), then?

:+1: for me, I'd like to hear how other feel about it.

---

One could argue that if we are to return the enclave name it should be extracted from the files (e.g. from permissions.xml) rather than inferred from location

Reply:
> One could argue that if we are to return the enclave name it should be extracted from the files (e.g. from permissions.xml) rather than inferred from location

Then that person should also be saying the lookup strategy in rcl is wrong :stuck_out_tongue: .

Reply:
:smile: for sure, I'll be honest and admit I'm not sure how rcl deals with it these days. I believe it takes the path with forward slashes and automagically resolves it the right way on windows.

It just seems less hacky to do "read value from file all files matching pattern" than "let's find the relative path of all the files matching that name but export it as a posix path and pretend it's absolute but as if the enclaves folder was the root".

On the other hand it's much easier to update the patch to use `as_posix()` and it fixes the current issue so I'm happy to do that as an incremental improvement :stuck_out_tongue:

Reply:
Haha, yeah I'm not sure about that part either, but it' still path-based, regardless of the slashes, so we might as well do the same thing here. Otherwise it could be pretty confusing to print the contents of the files but then have RCL not actually match those up on the backend, you know?

Reply:
fair enough :+1:

Reply:
I'm not sure where you're going, but I always saw the list command as a numerating a list of fully qualified enclave names. If the certificates were generated by SROS, then we could also look at the certificate common name, but those aren't guaranteed to match it's location as it can be moved in the key store. But I think we should stick with extrapolating the enclave name from its placement in the key store, as that's how RCL will resolve the enclave name within file system type key stores.

Reply:
> But I think we should stick with extrapolating the enclave name from its placement in the key store, as that's how RCL will resolve the enclave name within file system type key stores.

Sorry if I was unclear, that ^ is what I meant.

Reply:
added `as_posix` in https://github.com/ros2/sros2/pull/219/commits/23e41bf00a85fb7055c62b4428e6d81c945e3cec